File: develop_tbranchnumpy/test_performance.cxx
#include <iostream>
#include <ctime>
#include <sys/time.h>

#include "TFile.h"
#include "TTree.h"
#include "TH1D.h"
#include "TTreeReader.h"
#include "ROOT/TTreeReaderFast.hxx"
#include "ROOT/TTreeReaderValueFast.hxx"

const int WARM_UP = 5;
const int REPS = 100;

double diff(struct timeval endTime, struct timeval startTime) {
  return (1000L * 1000L * (endTime.tv_sec - startTime.tv_sec) + (endTime.tv_usec - startTime.tv_usec)) / 1000.0 / 1000.0;
}

double setBranchAddress_momentum(TTree *tree, int reps) {
  double total = 0.0;

  Float_t px;
  Float_t py;
  Float_t pz;
  tree->SetBranchAddress("px", &px);
  tree->SetBranchAddress("py", &py);
  tree->SetBranchAddress("pz", &pz);

  Long64_t numEntries = tree->GetEntries();

  for (int i = 0;  i < reps;  i++) {
    for (Long64_t entry = 0;  entry < numEntries;  entry++) {
      tree->GetEntry(entry);
      total += sqrt(px*px + py*py + pz*pz);
    }
  }

  return total;
}

double ttreeReaderFast_momentum(TTree *tree, int reps) {
  double total = 0.0;

  for (int i = 0;  i < 1;  i++) {
    ROOT::Experimental::TTreeReaderFast reader(tree);
    ROOT::Experimental::TTreeReaderValueFast<float> px(reader, "px");
    ROOT::Experimental::TTreeReaderValueFast<float> py(reader, "py");
    ROOT::Experimental::TTreeReaderValueFast<float> pz(reader, "pz");

    for (auto it = reader.begin();  it != reader.end();  ++it) {
        total += sqrt((*px)*(*px) + (*py)*(*py) + (*pz)*(*pz));
    }
  }

  return total;
}

double ttreeReader_momentum(TTree *tree, int reps) {
  double total = 0.0;

  TTreeReader reader(tree);
  TTreeReaderValue<float> px(reader, "px");
  TTreeReaderValue<float> py(reader, "py");
  TTreeReaderValue<float> pz(reader, "pz");

  for (int i = 0;  i < reps;  i++) {
    reader.Restart();
    while (reader.Next()) {
      total += sqrt((*px)*(*px) + (*py)*(*py) + (*pz)*(*pz));
    }
  }

  return total;
}

double ttreeDraw_momentum(TTree *tree, int reps) {
  TH1D total("total", "", 1, -10000.0, 10000.0);
  for (int i = 0;  i < reps;  i++) {
    tree->Draw("sqrt(px*px + py*py + pz*pz) >> total");
  }
  return total.GetBinContent(1);
}

double setBranchAddress_energy(TTree *tree, int reps) {
  double total = 0.0;

  Float_t px;
  Float_t py;
  Float_t pz;
  Float_t mass_mumu;
  tree->SetBranchAddress("px", &px);
  tree->SetBranchAddress("py", &py);
  tree->SetBranchAddress("pz", &pz);
  tree->SetBranchAddress("mass_mumu", &mass_mumu);

  Long64_t numEntries = tree->GetEntries();

  for (int i = 0;  i < reps;  i++) {
    for (Long64_t entry = 0;  entry < numEntries;  entry++) {
      tree->GetEntry(entry);
      total += sqrt(px*px + py*py + pz*pz + mass_mumu*mass_mumu);
    }
  }

  return total;
}

double ttreeReaderFast_energy(TTree *tree, int reps) {
  double total = 0.0;

  for (int i = 0;  i < 1;  i++) {
    ROOT::Experimental::TTreeReaderFast reader(tree);
    ROOT::Experimental::TTreeReaderValueFast<float> px(reader, "px");
    ROOT::Experimental::TTreeReaderValueFast<float> py(reader, "py");
    ROOT::Experimental::TTreeReaderValueFast<float> pz(reader, "pz");
    ROOT::Experimental::TTreeReaderValueFast<float> mass_mumu(reader, "mass_mumu");

    for (auto it = reader.begin();  it != reader.end();  ++it) {
        total += sqrt((*px)*(*px) + (*py)*(*py) + (*pz)*(*pz) + (*mass_mumu)*(*mass_mumu));
    }
  }

  return total;
}

double ttreeReader_energy(TTree *tree, int reps) {
  double total = 0.0;

  TTreeReader reader(tree);
  TTreeReaderValue<float> px(reader, "px");
  TTreeReaderValue<float> py(reader, "py");
  TTreeReaderValue<float> pz(reader, "pz");
  TTreeReaderValue<float> mass_mumu(reader, "mass_mumu");

  for (int i = 0;  i < reps;  i++) {
    reader.Restart();
    while (reader.Next()) {
      total += sqrt((*px)*(*px) + (*py)*(*py) + (*pz)*(*pz) + (*mass_mumu)*(*mass_mumu));
    }
  }

  return total;
}

double ttreeDraw_energy(TTree *tree, int reps) {
  TH1D total("total", "", 1, -10000.0, 10000.0);
  for (int i = 0;  i < reps;  i++) {
    tree->Draw("sqrt(px*px + py*py + pz*pz + mass_mumu*mass_mumu) >> total");
  }
  return total.GetBinContent(1);
}

void test_performance() {
  {
    TFile* file = TFile::Open("TrackResonanceNtuple_uncompressed.root");
    TTree* tree;
    file->GetObject("twoMuon", tree);
    tree->SetBranchStatus("*", 0);
    tree->SetBranchStatus("px", 1);
    tree->SetBranchStatus("py", 1);
    tree->SetBranchStatus("pz", 1);

    setBranchAddress_momentum(tree, WARM_UP);

    struct timeval startTime, endTime;
    gettimeofday(&startTime, 0);
    setBranchAddress_momentum(tree, REPS);
    gettimeofday(&endTime, 0);

    std::cout << "TrackResonanceNtuple_uncompressed.root momentum SetBranchAddress " << diff(endTime, startTime) << " sec" << std::endl;
  }

  {
    TFile* file = TFile::Open("TrackResonanceNtuple_compressed.root");
    TTree* tree;
    file->GetObject("twoMuon", tree);
    tree->SetBranchStatus("*", 0);
    tree->SetBranchStatus("px", 1);
    tree->SetBranchStatus("py", 1);
    tree->SetBranchStatus("pz", 1);

    setBranchAddress_momentum(tree, WARM_UP);

    struct timeval startTime, endTime;
    gettimeofday(&startTime, 0);
    setBranchAddress_momentum(tree, REPS);
    gettimeofday(&endTime, 0);

    std::cout << "TrackResonanceNtuple_compressed.root momentum SetBranchAddress " << diff(endTime, startTime) << " sec" << std::endl;
  }

  {
    TFile* file = TFile::Open("TrackResonanceNtuple_uncompressed.root");
    TTree* tree;
    file->GetObject("twoMuon", tree);
    tree->SetBranchStatus("*", 0);
    tree->SetBranchStatus("px", 1);
    tree->SetBranchStatus("py", 1);
    tree->SetBranchStatus("pz", 1);

    ttreeReader_momentum(tree, WARM_UP);

    struct timeval startTime, endTime;
    gettimeofday(&startTime, 0);
    ttreeReader_momentum(tree, REPS);
    gettimeofday(&endTime, 0);

    std::cout << "TrackResonanceNtuple_uncompressed.root momentum TTreeReader " << diff(endTime, startTime) << " sec" << std::endl;
  }

  {
    TFile* file = TFile::Open("TrackResonanceNtuple_compressed.root");
    TTree* tree;
    file->GetObject("twoMuon", tree);
    tree->SetBranchStatus("*", 0);
    tree->SetBranchStatus("px", 1);
    tree->SetBranchStatus("py", 1);
    tree->SetBranchStatus("pz", 1);

    ttreeReader_momentum(tree, WARM_UP);

    struct timeval startTime, endTime;
    gettimeofday(&startTime, 0);
    ttreeReader_momentum(tree, REPS);
    gettimeofday(&endTime, 0);

    std::cout << "TrackResonanceNtuple_compressed.root momentum TTreeReader " << diff(endTime, startTime) << " sec" << std::endl;
  }

  {
    TFile* file = TFile::Open("TrackResonanceNtuple_uncompressed.root");
    TTree* tree;
    file->GetObject("twoMuon", tree);
    tree->SetBranchStatus("*", 0);
    tree->SetBranchStatus("px", 1);
    tree->SetBranchStatus("py", 1);
    tree->SetBranchStatus("pz", 1);

    ttreeDraw_momentum(tree, WARM_UP);

    struct timeval startTime, endTime;
    gettimeofday(&startTime, 0);
    ttreeDraw_momentum(tree, REPS);
    gettimeofday(&endTime, 0);

    std::cout << "TrackResonanceNtuple_uncompressed.root momentum TTree::Draw " << diff(endTime, startTime) << " sec" << std::endl;
  }

  {
    TFile* file = TFile::Open("TrackResonanceNtuple_compressed.root");
    TTree* tree;
    file->GetObject("twoMuon", tree);
    tree->SetBranchStatus("*", 0);
    tree->SetBranchStatus("px", 1);
    tree->SetBranchStatus("py", 1);
    tree->SetBranchStatus("pz", 1);

    ttreeDraw_momentum(tree, WARM_UP);

    struct timeval startTime, endTime;
    gettimeofday(&startTime, 0);
    ttreeDraw_momentum(tree, REPS);
    gettimeofday(&endTime, 0);

    std::cout << "TrackResonanceNtuple_compressed.root momentum TTree::Draw " << diff(endTime, startTime) << " sec" << std::endl;
  }

  // {
  //   TFile* file = TFile::Open("TrackResonanceNtuple_uncompressed.root");
  //   TTree* tree;
  //   file->GetObject("twoMuon", tree);
  //   tree->SetBranchStatus("*", 0);
  //   tree->SetBranchStatus("px", 1);
  //   tree->SetBranchStatus("py", 1);
  //   tree->SetBranchStatus("pz", 1);

  //   ttreeReaderFast_momentum(tree, WARM_UP);

  //   struct timeval startTime, endTime;
  //   gettimeofday(&startTime, 0);
  //   ttreeReaderFast_momentum(tree, REPS);
  //   gettimeofday(&endTime, 0);

  //   std::cout << "TrackResonanceNtuple_uncompressed.root momentum TTreeReaderFast " << diff(endTime, startTime) << " sec" << std::endl;
  // }

  // {
  //   TFile* file = TFile::Open("TrackResonanceNtuple_compressed.root");
  //   TTree* tree;
  //   file->GetObject("twoMuon", tree);
  //   tree->SetBranchStatus("*", 0);
  //   tree->SetBranchStatus("px", 1);
  //   tree->SetBranchStatus("py", 1);
  //   tree->SetBranchStatus("pz", 1);

  //   ttreeReaderFast_momentum(tree, WARM_UP);

  //   struct timeval startTime, endTime;
  //   gettimeofday(&startTime, 0);
  //   ttreeReaderFast_momentum(tree, REPS);
  //   gettimeofday(&endTime, 0);

  //   std::cout << "TrackResonanceNtuple_compressed.root momentum TTreeReaderFast " << diff(endTime, startTime) << " sec" << std::endl;
  // }

  {
    TFile* file = TFile::Open("TrackResonanceNtuple_uncompressed.root");
    TTree* tree;
    file->GetObject("twoMuon", tree);
    tree->SetBranchStatus("*", 0);
    tree->SetBranchStatus("px", 1);
    tree->SetBranchStatus("py", 1);
    tree->SetBranchStatus("pz", 1);
    tree->SetBranchStatus("mass_mumu", 1);

    setBranchAddress_energy(tree, WARM_UP);

    struct timeval startTime, endTime;
    gettimeofday(&startTime, 0);
    setBranchAddress_energy(tree, REPS);
    gettimeofday(&endTime, 0);

    std::cout << "TrackResonanceNtuple_uncompressed.root energy SetBranchAddress " << diff(endTime, startTime) << " sec" << std::endl;
  }

  {
    TFile* file = TFile::Open("TrackResonanceNtuple_compressed.root");
    TTree* tree;
    file->GetObject("twoMuon", tree);
    tree->SetBranchStatus("*", 0);
    tree->SetBranchStatus("px", 1);
    tree->SetBranchStatus("py", 1);
    tree->SetBranchStatus("pz", 1);
    tree->SetBranchStatus("mass_mumu", 1);

    setBranchAddress_energy(tree, WARM_UP);

    struct timeval startTime, endTime;
    gettimeofday(&startTime, 0);
    setBranchAddress_energy(tree, REPS);
    gettimeofday(&endTime, 0);

    std::cout << "TrackResonanceNtuple_compressed.root energy SetBranchAddress " << diff(endTime, startTime) << " sec" << std::endl;
  }

  {
    TFile* file = TFile::Open("TrackResonanceNtuple_uncompressed.root");
    TTree* tree;
    file->GetObject("twoMuon", tree);
    tree->SetBranchStatus("*", 0);
    tree->SetBranchStatus("px", 1);
    tree->SetBranchStatus("py", 1);
    tree->SetBranchStatus("pz", 1);
    tree->SetBranchStatus("mass_mumu", 1);

    ttreeReader_energy(tree, WARM_UP);

    struct timeval startTime, endTime;
    gettimeofday(&startTime, 0);
    ttreeReader_energy(tree, REPS);
    gettimeofday(&endTime, 0);

    std::cout << "TrackResonanceNtuple_uncompressed.root energy TTreeReader " << diff(endTime, startTime) << " sec" << std::endl;
  }

  {
    TFile* file = TFile::Open("TrackResonanceNtuple_compressed.root");
    TTree* tree;
    file->GetObject("twoMuon", tree);
    tree->SetBranchStatus("*", 0);
    tree->SetBranchStatus("px", 1);
    tree->SetBranchStatus("py", 1);
    tree->SetBranchStatus("pz", 1);
    tree->SetBranchStatus("mass_mumu", 1);

    ttreeReader_energy(tree, WARM_UP);

    struct timeval startTime, endTime;
    gettimeofday(&startTime, 0);
    ttreeReader_energy(tree, REPS);
    gettimeofday(&endTime, 0);

    std::cout << "TrackResonanceNtuple_compressed.root energy TTreeReader " << diff(endTime, startTime) << " sec" << std::endl;
  }

  {
    TFile* file = TFile::Open("TrackResonanceNtuple_uncompressed.root");
    TTree* tree;
    file->GetObject("twoMuon", tree);
    tree->SetBranchStatus("*", 0);
    tree->SetBranchStatus("px", 1);
    tree->SetBranchStatus("py", 1);
    tree->SetBranchStatus("pz", 1);
    tree->SetBranchStatus("mass_mumu", 1);

    ttreeDraw_energy(tree, WARM_UP);

    struct timeval startTime, endTime;
    gettimeofday(&startTime, 0);
    ttreeDraw_energy(tree, REPS);
    gettimeofday(&endTime, 0);

    std::cout << "TrackResonanceNtuple_uncompressed.root energy TTree::Draw " << diff(endTime, startTime) << " sec" << std::endl;
  }

  {
    TFile* file = TFile::Open("TrackResonanceNtuple_compressed.root");
    TTree* tree;
    file->GetObject("twoMuon", tree);
    tree->SetBranchStatus("*", 0);
    tree->SetBranchStatus("px", 1);
    tree->SetBranchStatus("py", 1);
    tree->SetBranchStatus("pz", 1);
    tree->SetBranchStatus("mass_mumu", 1);

    ttreeDraw_energy(tree, WARM_UP);

    struct timeval startTime, endTime;
    gettimeofday(&startTime, 0);
    ttreeDraw_energy(tree, REPS);
    gettimeofday(&endTime, 0);

    std::cout << "TrackResonanceNtuple_compressed.root energy TTree::Draw " << diff(endTime, startTime) << " sec" << std::endl;
  }

  // {
  //   TFile* file = TFile::Open("TrackResonanceNtuple_uncompressed.root");
  //   TTree* tree;
  //   file->GetObject("twoMuon", tree);
  //   tree->SetBranchStatus("*", 0);
  //   tree->SetBranchStatus("px", 1);
  //   tree->SetBranchStatus("py", 1);
  //   tree->SetBranchStatus("pz", 1);
  //   tree->SetBranchStatus("mass_mumu", 1);

  //   ttreeReaderFast_energy(tree, WARM_UP);

  //   struct timeval startTime, endTime;
  //   gettimeofday(&startTime, 0);
  //   ttreeReaderFast_energy(tree, REPS);
  //   gettimeofday(&endTime, 0);

  //   std::cout << "TrackResonanceNtuple_uncompressed.root energy TTreeReaderFast " << diff(endTime, startTime) << " sec" << std::endl;
  // }

  // {
  //   TFile* file = TFile::Open("TrackResonanceNtuple_compressed.root");
  //   TTree* tree;
  //   file->GetObject("twoMuon", tree);
  //   tree->SetBranchStatus("*", 0);
  //   tree->SetBranchStatus("px", 1);
  //   tree->SetBranchStatus("py", 1);
  //   tree->SetBranchStatus("pz", 1);
  //   tree->SetBranchStatus("mass_mumu", 1);

  //   ttreeReaderFast_energy(tree, WARM_UP);

  //   struct timeval startTime, endTime;
  //   gettimeofday(&startTime, 0);
  //   ttreeReaderFast_energy(tree, REPS);
  //   gettimeofday(&endTime, 0);

  //   std::cout << "TrackResonanceNtuple_compressed.root energy TTreeReaderFast " << diff(endTime, startTime) << " sec" << std::endl;
  // }

}

File: develop_tbranchnumpy/test_performance.py
import math
import time

import numpy
import numba
import ROOT
import root_numpy

import numpyinterface

@numba.jit(nopython=True)
def momentumsum(px, py, pz):
    total = 0.0
    for i in range(len(px)):
        total += math.sqrt(px[i]**2 + py[i]**2 + pz[i]**2)
    return total

@numba.jit(nopython=True)
def energysum(px, py, pz, mass_mumu):
    total = 0.0
    for i in range(len(px)):
        total += math.sqrt(px[i]**2 + py[i]**2 + pz[i]**2 + mass_mumu[i]**2)
    return total

def numpy_momentum(reps, fileName, return_new_buffers, swap_bytes):
    iterator = numpyinterface.iterate(
        fileName, "twoMuon", "px", "py", "pz",
        return_new_buffers = return_new_buffers,
        swap_bytes = swap_bytes)

    startTime = time.time()

    for i in xrange(reps):
        total = 0
        for start, end, px, py, pz in iterator:
            total = numpy.sqrt(px**2 + py**2 + pz**2).sum()

    endTime = time.time()

    return endTime - startTime

def numba_momentum(reps, fileName, return_new_buffers, swap_bytes):
    iterator = numpyinterface.iterate(
        fileName, "twoMuon", "px", "py", "pz",
        return_new_buffers = return_new_buffers,
        swap_bytes = swap_bytes)

    startTime = time.time()

    for i in xrange(reps):
        total = 0
        for start, end, px, py, pz in iterator:
            total = momentumsum(px, py, pz)

    endTime = time.time()

    return endTime - startTime

def rootnumpy_momentum(reps, fileName):
    file = ROOT.TFile(fileName)
    tree = file.Get("twoMuon")

    startTime = time.time()

    for i in xrange(reps):
        array = root_numpy.tree2array(tree, branches=["px", "py", "pz"])
        px = array["px"]
        py = array["py"]
        pz = array["pz"]
        total = numpy.sqrt(px**2 + py**2 + pz**2).sum()

    endTime = time.time()

    return endTime - startTime

def numpy_energy(reps, fileName, return_new_buffers, swap_bytes):
    iterator = numpyinterface.iterate(
        fileName, "twoMuon", "px", "py", "pz", "mass_mumu",
        return_new_buffers = return_new_buffers,
        swap_bytes = swap_bytes)

    startTime = time.time()

    for i in xrange(reps):
        total = 0
        for start, end, px, py, pz, mass_mumu in iterator:
            total = numpy.sqrt(px**2 + py**2 + pz**2 + mass_mumu**2).sum()

    endTime = time.time()

    return endTime - startTime

def numba_energy(reps, fileName, return_new_buffers, swap_bytes):
    iterator = numpyinterface.iterate(
        fileName, "twoMuon", "px", "py", "pz", "mass_mumu",
        return_new_buffers = return_new_buffers,
        swap_bytes = swap_bytes)

    startTime = time.time()

    for i in xrange(reps):
        total = 0
        for start, end, px, py, pz, mass_mumu in iterator:
            total = energysum(px, py, pz, mass_mumu)

    endTime = time.time()

    return endTime - startTime

def rootnumpy_energy(reps, fileName):
    file = ROOT.TFile(fileName)
    tree = file.Get("twoMuon")

    startTime = time.time()

    for i in xrange(reps):
        array = root_numpy.tree2array(tree, branches=["px", "py", "pz", "mass_mumu"])
        px = array["px"]
        py = array["py"]
        pz = array["pz"]
        mass_mumu = array["mass_mumu"]
        total = numpy.sqrt(px**2 + py**2 + pz**2 + mass_mumu**2).sum()

    endTime = time.time()

    return endTime - startTime

WARM_UP = 5
REPS = 100

for fileName in "TrackResonanceNtuple_uncompressed.root", "TrackResonanceNtuple_compressed.root":
    for label, return_new_buffers in ("new buffers", True), ("views", False):
        print fileName, "momentum numpy big-endian", label,
        numpy_momentum(WARM_UP, fileName, return_new_buffers, False)     # warm up
        print numpy_momentum(REPS, fileName, return_new_buffers, False)  # real run

        print fileName, "momentum numpy little-endian", label,
        numpy_momentum(WARM_UP, fileName, return_new_buffers, True)      # warm up
        print numpy_momentum(REPS, fileName, return_new_buffers, True)   # real run

        print fileName, "momentum numba little-endian", label,
        numba_momentum(WARM_UP, fileName, return_new_buffers, True)      # warm up
        print numba_momentum(REPS, fileName, return_new_buffers, True)   # real run

    print fileName, "momentum root_numpy",
    rootnumpy_momentum(WARM_UP, fileName)                                # warm up
    print rootnumpy_momentum(REPS, fileName)                             # real run

print
print numpyinterface.performance()
print

for fileName in "TrackResonanceNtuple_uncompressed.root", "TrackResonanceNtuple_compressed.root":
    for label, return_new_buffers in ("new buffers", True), ("views", False):
        print fileName, "energy numpy big-endian", label,
        numpy_energy(WARM_UP, fileName, return_new_buffers, False)       # warm up
        print numpy_energy(REPS, fileName, return_new_buffers, False)    # real run

        print fileName, "energy numpy little-endian", label,
        numpy_energy(WARM_UP, fileName, return_new_buffers, True)        # warm up
        print numpy_energy(REPS, fileName, return_new_buffers, True)     # real run

        print fileName, "energy numba little-endian", label,
        numba_energy(WARM_UP, fileName, return_new_buffers, True)        # warm up
        print numba_energy(REPS, fileName, return_new_buffers, True)     # real run

    print fileName, "energy root_numpy",
    rootnumpy_energy(WARM_UP, fileName)                                  # warm up
    print rootnumpy_energy(REPS, fileName)                               # real run

print
print numpyinterface.performance()
print

